### PR TITLE
Ensure we not sent duplicate certificates when using OpenSslEngine

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We need to ensure we not set duplicated certificates when using OpenSslEngine.

Modifications:

- Skip first cert in chain when set the chain itself and so not send duplicated certificates
- Add interopt unit tests to ensure no duplicates are send.

Result:

No more duplicates.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@carl-mastrangelo FYI... This is related to https://github.com/netty/netty-tcnative/issues/184

@nmittler @Scottmitch PTAL

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 1]
```diff
@@ -1857,10 +1849,10 @@ private String selectApplicationProtocol(List<String> protocols,
         @Override
         public Certificate[] getPeerCertificates() throws SSLPeerUnverifiedException {
             synchronized (ReferenceCountedOpenSslEngine.this) {
-                if (peerCerts == null || peerCerts.length == 0) {
+                if (peerCerts == null || peerCerts.isEmpty()) {
                     throw new SSLPeerUnverifiedException("peer not verified");
                 }
-                return peerCerts;
+                return peerCerts.toArray(new Certificate[peerCerts.size()]);
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: can we just clone the array instead and stick with the array?

Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: @Scottmitch why ? I mean I can do... but why ?

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java [hunk 2]
```diff
@@ -236,15 +276,15 @@ private void mySetupMutualAuth(
...
+                .sslProvider(sslServerProvider())
                 .trustManager(servertTrustCrtFile)
                 .ciphers(null, IdentityCipherSuiteFilter.INSTANCE)
                 .sessionCacheSize(0)
                 .sessionTimeout(0)
                 .build();
 
         clientSslCtx = SslContextBuilder.forClient()
-                .sslProvider(sslProvider())
+                .sslProvider(sslClientProvider())
```
Comment (Scottmitch) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java]: `cleanupSslContext` assumed both context are generated using the same sslProvider. Consider breaking this into 2 methods as well.

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 1 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: we should know the exact size right? It was an array before ... I'm wondering the same thing "why" :)

Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: @Scottmitch it just makes the `init*` method a lot easier to use an ArrayList. That said I can change if you want.

Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: would this change require the init code to become much more complicated? I prefer to keep it if its not significantly more complicated ... better to just allocate what we need and we don't expect it to grow after its created ... seems like a good use case for a fixed size array

Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: @Scottmitch let me see...

**Comment by normanmaurer** [comment]
@Scottmitch PTAL again... addressed everything

**Comment by Scottmitch** [comment]
lgtm

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (af63227) and 4.0 (4fee26d7be1556c00af817315c1d174f927ca21b)
